internal/protocol/connector.go

```diff
@@ -236,17 +238,30 @@ func Handshake(ctx context.Context, conn net.Conn, version uint64) (*Protocol, e
 		conn.SetDeadline(deadline)
 		defer conn.SetDeadline(time.Time{})
 	}
+	// Honor context cancellation.
+	canceled := false
+	stop := context.AfterFunc(ctx, func() {
+		if ctx.Err() == context.Canceled {
+			canceled = true
+			// Cancel read and writes by setting deadline in the past.
+			conn.SetDeadline(time.Unix(1, 0))
+		}
+	})
+	defer stop()
 
 	// Perform the protocol handshake.
 	n, err := conn.Write(protocol)
 	if err != nil {
+		if canceled && errors.Is(err, os.ErrDeadlineExceeded) {
```

Comment:
`canceled` is being written from one goroutine and read from another, is there a race?

Reply:
Also why do we have a separate `if` for this case when it's handled identically to the one just below?

Reply:
Changed the error message to say canceled, this was more for prepping for the next change but it makes sense to just add it now. Thanks for spotting it!